File: src/main/java/com/mindscapehq/android/raygun4android/network/http/RaygunHttpUrlStreamHandler.java
package main.java.com.mindscapehq.android.raygun4android.network.http;

import main.java.com.mindscapehq.android.raygun4android.RaygunLogger;
import main.java.com.mindscapehq.android.raygun4android.network.RaygunNetworkUtils;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.Proxy;
import java.net.URL;
import java.net.URLConnection;
import java.net.URLStreamHandler;


public final class RaygunHttpUrlStreamHandler extends URLStreamHandler
{
  private static final int PORT = 80;
  private static final String PROTOCOL = "http";
  private URLStreamHandler originalHandler;

  public RaygunHttpUrlStreamHandler(URLStreamHandler handler)
  {
    originalHandler = handler;
  }

  protected URLConnection openConnection(URL url)
      throws IOException
  {
    try
    {
      Method method = RaygunNetworkUtils.findMethod(originalHandler.getClass(), "openConnection", new Class<?>[]{ URL.class });
      method.setAccessible(true);

      URLConnection urlConnection = (URLConnection) method.invoke(originalHandler, url);

      if (urlConnection == null) {
        throw new IOException("Failed to create connection");
      }

      return new RaygunHttpUrlConnection(urlConnection);
    }
    catch(NoSuchMethodException e) {
      RaygunLogger.e("Exception occurred in openConnection: " + e.getMessage());
    }
    catch(IllegalAccessException e) {
      RaygunLogger.e("Exception occurred in openConnection: " + e.getMessage());
    }
    catch(InvocationTargetException e) {
      RaygunLogger.e("Exception occurred in openConnection: " + e.getMessage());
    }

    throw new IOException();
  }

  protected URLConnection openConnection(URL url, Proxy proxy)
      throws IOException
  {
    try
    {
      Method method = RaygunNetworkUtils.findMethod(originalHandler.getClass(), "openConnection", new Class<?>[]{ URL.class, Proxy.class });
      method.setAccessible(true);

      URLConnection urlConnection = (URLConnection) method.invoke(originalHandler, url, proxy);

      if (urlConnection == null) {
        throw new IOException("Failed to create connection");
      }

      return new RaygunHttpUrlConnection(urlConnection);
    }
    catch(NoSuchMethodException e) {
      RaygunLogger.e("Exception occurred in openConnection: " + e.getMessage());
    }
    catch(IllegalAccessException e) {
      RaygunLogger.e("Exception occurred in openConnection: " + e.getMessage());
    }
    catch(InvocationTargetException e) {
      RaygunLogger.e("Exception occurred in openConnection: " + e.getMessage());
    }

    throw new IOException();
  }

  public int getDefaultPort()
  {
    return PORT;
  }

  public String getProtocol()
  {
    return PROTOCOL;
  }
}

File: provider/src/main/java/com/raygun/raygun4android/network/http/RaygunHttpUrlStreamHandler.kt
package com.raygun.raygun4android.network.http

import com.raygun.raygun4android.logging.RaygunLogger.e
import com.raygun.raygun4android.utils.RaygunReflectionUtils
import java.io.IOException
import java.lang.reflect.InvocationTargetException
import java.net.Proxy
import java.net.URL
import java.net.URLConnection
import java.net.URLStreamHandler

internal class RaygunHttpUrlStreamHandler(private val originalHandler: URLStreamHandler) :
    URLStreamHandler() {
    @Throws(IOException::class)
    override fun openConnection(url: URL): URLConnection {
        try {
            val method =
                RaygunReflectionUtils.findMethod(
                    originalHandler.javaClass,
                    "openConnection",
                    arrayOf<Class<*>>(URL::class.java)
                )
            method.isAccessible = true

            val urlConnection = method.invoke(originalHandler, url) as URLConnection
                ?: throw IOException("Failed to create connection")

            return RaygunHttpUrlConnection(urlConnection)
        } catch (e: NoSuchMethodException) {
            e("Exception occurred in openConnection: " + e.message)
        } catch (e: IllegalAccessException) {
            e("Exception occurred in openConnection: " + e.message)
        } catch (e: InvocationTargetException) {
            e("Exception occurred in openConnection: " + e.message)
        }

        throw IOException()
    }

    @Throws(IOException::class)
    override fun openConnection(url: URL, proxy: Proxy): URLConnection {
        try {
            val method =
                RaygunReflectionUtils.findMethod(
                    originalHandler.javaClass,
                    "openConnection",
                    arrayOf(URL::class.java, Proxy::class.java)
                )
            method.isAccessible = true

            val urlConnection =
                method.invoke(originalHandler, url, proxy) as URLConnection
                    ?: throw IOException("Failed to create connection")

            return RaygunHttpUrlConnection(urlConnection)
        } catch (e: NoSuchMethodException) {
            e("Exception occurred in openConnection: " + e.message)
        } catch (e: IllegalAccessException) {
            e("Exception occurred in openConnection: " + e.message)
        } catch (e: InvocationTargetException) {
            e("Exception occurred in openConnection: " + e.message)
        }

        throw IOException()
    }

    public override fun getDefaultPort(): Int {
        return PORT
    }

    companion object {
        private const val PORT = 80
        const val PROTOCOL = "http"
    }
}
